[리액트 8기 이재윤] 8주차 과제 1

jaeyooniverse opened the pull request:
마지막 주도 잘 부탁드립니다!

jaeyooniverse committed:
첫 번째 강의 내용 업데이트

jaeyooniverse committed:
두 번째 강의 내용 업데이트

jaeyooniverse committed:
세 번째 강의 내용 업데이트

jaeyooniverse committed:
한 파일에는 하나의 컴포넌트만 있어야 한다는 요구 사항을 충족하기 위한 파일 분리

한 파일에는 하나의 컴포넌트만 존재해야 하므로 상태 관심사의 containers 디렉토리와 UI 관심사의 pages 디렉토리
또한 UI 관심사에서 MenuList, MenuItems, TextField와 같이 공통으로 사용하는 컴포넌트들은 pages안 common 디렉토리를
추가하여 분리한다. redux 관련 파일들은 redux 폴더를 만들어 넣어준다.

jaeyooniverse commented:
먼저, 오늘은 강의 내용을 업데이트하고 한 파일에는 하나의 컴포넌트만 있어야 한다는 요구 사항을 충족시키기 위한 
파일 분리까지 진행했습니다.

과제를 하면서 궁금한 점이 몇 가지 있습니다.

1.
파일 분리는 팀의 컨벤션에 따라 나눈다는 리뷰를 받은 적이 있었는데, redux 관련 파일들은 redux라는 폴더를 만들어 넣어주고, pages에서 공통으로 쓰이는 컴포넌트는 commons라는 폴더를 만들어 넣어줬습니다. 크게 문제없이 잘한걸까요?

2.
Redux toolkit을 사용하여 중복 코드 제거하기는 강의를 통해 끝난 것이 아닌가요? 아니면 더 제거할 수 있는 코드가 있는지 궁금합니다.

3.
Emotion을 이용하여 꾸미기는... 공대생이라 미적 감각이 많이 부족합니다. 윤석님의 너른 양해 부탁드립니다.

4.
UseCallback과 React.memo는 모든 컴포넌트에 적용하라는 뜻인건가요? 아직 PureComponent에 대한 이해가 부족해서 Virtaul DOM의 렌더링 방식이 어떻게 다른지 공부해야겠다는 필요성을 느꼈습니다...!

항상 감사합니다!

jaeyooniverse committed:
index.jsx의 store 경로 수정

index.jsx의 경로가 잘못되어 있어서 이를 수정했습니다.

hannut91 commented:
> 파일 분리는 팀의 컨벤션에 따라 나눈다는 리뷰를 받은 적이 있었는데, redux 관련 파일들은 redux라는 폴더를 만들어 넣어주고, pages에서 공통으로 쓰이는 컴포넌트는 commons라는 폴더를 만들어 넣어줬습니다. 크게 문제없이 잘한걸까요?

음... 진짜 이 문제는 정답이 없는거라서 뭔가 조언을 드리기가 어려운데요. 공통으로 사용되는 것들은 `shared`라고 많이 쓰는 것 같긴해요. 

https://reactjs.org/docs/faq-structure.html

위에서 얘기한 것처럼 엄청 중요하진 않고 일관성있게 유지하는게 좋은 것 같아요. (팀원들과 합의된 방향으로)

> Redux toolkit을 사용하여 중복 코드 제거하기는 강의를 통해 끝난 것이 아닌가요? 아니면 더 제거할 수 있는 코드가 있는지 궁금합니다.

강의에서 배운 것을 직접 해보세요 ㅎㅎ

> Emotion을 이용하여 꾸미기는... 공대생이라 미적 감각이 많이 부족합니다. 윤석님의 너른 양해 부탁드립니다.

많이 따라하다보면 늡니다 ㅎㅎ 아래 영상을 보면서 따라해 보세요.

https://www.youtube.com/c/OnlineTutorials4Designers

> UseCallback과 React.memo는 모든 컴포넌트에 적용하라는 뜻인건가요? 아직 PureComponent에 대한 이해가 부족해서 Virtaul DOM의 렌더링 방식이 어떻게 다른지 공부해야겠다는 필요성을 느꼈습니다...!

적용할 수 있는 모든 컴포넌트에 적용해 봅시다. 연습해 보는 거니까요. 실무에서는 적절하게 선택하는게 중요합니다.

리액트에서 컴포넌트는 함수입니다. 순수 함수가 무엇인지에 대해서 고민해 보시고, 컴포넌트 관점에서 순수 하다는 것은 무엇인지 생각해 보세요. 아마 2주차 때 강의 때 배웠던 것을 떠올려보시면 도움이 되실 것 같아요.

hannut91 commented on src/redux/slice.js:
```diff
@@ -0,0 +1,229 @@
+      reviewFields: { score, description },
+    } = getState();
+
+    await postReview({
+      accessToken,
+      restaurantId,
+      score,
+      description,
+    });
```
에러 처리 연습!

jaeyooniverse replied:
이번 주는 기필코 반드시! 💪

jaeyooniverse committed:
성능 최적화를 위한 React.memo() 사용

함수형 컴포넌트에서 props의 변경이 잘 일어나지 않는 컴포넌트들에 성능 향상을 위한 React.memo()를 적용하였습니다.

jaeyooniverse commented:
성능 최적화를 위한 React.memo()를 적용해봤습니다.

처음부터 모든 컴포넌트에 React.memo()를 적용하는 것 보다 React.memo()는 함수형 컴포넌트에서 props의 변경이 잘 일어나지 않는 컴포넌트에 성능 향상을 위한 기능이므로

1. 함수형 컴포넌트인가?
2. 해당 컴포넌트가 받는 prop은 자주 변경되지 않는 prop인가?

라는 저만의 기준으로 모두 통과하면 React.memo()를 적용해봤습니다.

React.memo()를 적용해보면서 React.memo()를 사용하기 위한 어떤 명확한 기준이 있으면 좋을 것 같은데, 사용자의 판단에 맡겨서 그 부분이 쉽지 않았다는 느낌을 받았고, 강의 영상에서는 Reviews에 React.memo()를 적용했지만, Reviews는 말 그대로 사용자가 리뷰를 작성하면 prop이 자주 변경되니까 React.memo()를 제거했습니다.

https://github.com/CodeSoom/react-week8-assignment-1/pull/75/commits/22b5cb4e36a451ac9ede4d041f7d70365b796fb2 해당 커밋에 이 컴포넌트는 React.memo()를 사용해도 되지 않을까? 반대로 이 컴포넌트는 React.memo()를 사용하면 안될 것 같은데와 같이 제가 잘못한 부분이 있으면 편하게 리뷰 부탁드립니다. 감사합니다!

jaeyooniverse committed:
커버리지 100% 달성을 위한 actions.test.js 추가

부족한 slice.test.js 커버리지를 보충하기 위한 테스트 코드를 추가했습니다.

jaeyooniverse committed:
index.jsx에서 store의 경로가 slice에서 store로 수정

기존 index.jsx의 store를 store가 아닌 slice로 경로가 설정되어 있어서 에러가 발생하였기에
이를 수정했습니다.

jaeyooniverse committed:
불필요한 시간 낭비를 줄이기 위한 Promise 설정

기존 코드에서 loadInitialData 액션 크리에이터 함수는 지역 정보를 가져오는 시간, 카테고리 정보를 가져오는 시간
동시에 진행하는 것이 아닌 지역 정보를 가져온 다음, 카테고리 정보를 가져오는 방식으로 진행되었는데,
이를 Promise를 통해 비동기로 동시에 진행하도록 수정했습니다.

jaeyooniverse committed:
반복되는 Link, Item 제거

반복되는 Link와 Item을 객체로 만들어 map을 통해 반복을 줄였습니다.

hannut91 commented:
`React.memo()`를 올바르게 사용하는 방법은 명확하게 아셔야 합니다. 잘못사용하면 올바르게 동작하지 않아서 사용하지 않는 것보다 더 안 좋아질 수 있어요. 예를들어서 `React.memo`를 사용했는데 props로 익명함수를 받아서 매번 값이 달라진다고 판단하여 다시 렌더링하고 다시 메모이제이션을 하면 오히려 안하는 것보다 낫겠죠. 이렇게 제대로 사용하는 방법은 완전히 익히셔야 합니다.

근데 이 `React.memo`를 사용해야 하는 때는 명확한 기준은 없습니다. 그냥 명목적으로 적용하면 구현하는데 시간만 더 걸리고 복잡해지기만 합니다. 그래서 `측정`이 중요합니다. 성능을 측정하고, 성능을 얼마나 개선할지 목표를 정하고 실제로 개선한 후 다시 측정해서 실제로 효과가 있는지를 확인해야 하는게 정석적인 방법입니다.

또 생각해볼점은 처음부터 성능을 고려하는 것은 별로 좋지 않습니다. 왜냐하면 너무 이른 최적화는 구현을 복잡하게 하기 때문입니다. 아주 필요한 경우에만 하는게 좋습니다. 지금은 연습하는 과정이니 사용하는 법을 충분히 익혀두고, 나중에 필요할 때 적절히 사용해야 합니다. 어떻게 측정해야 하는지도 공부해 보세요. 

https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en

hannut91 commented:
승민님에게 얘기하고 싶었던 것은 RTK Query를 사용하라는 것이었어요 ㅎㅎ

https://redux-toolkit.js.org/rtk-query/overview

jaeyooniverse committed:
에러 처리를 위한 reducer 분리

에러 처리를 위해 redux에 있는 state와 reducer를 종류별로 나누고 각각의 state에 isLoading과
isError를 추가하여 에러 핸들링 기능을 도입하고자 추가했습니다.

jaeyooniverse committed:
reducer 분리 후 에러 핸들링 테스트 롤백

reducer를 분리 한 다음에 에러 핸들링 테스트를 진행해보려 했으나 원하는 결과를 얻지 못해서 롤백했습니다.

jaeyooniverse commented:
생에 처음으로 CSS를 제대로 하려니까 어떻게 기준을 잡고 시작해야 할 지 너무 막막하네요.
예전 간단한 홈페이지를 만들 때도 UI가 어떻게 그려질지 전혀 예상이 되지 않는 채로 수치를 막 입력한 적이 있었는데,
예측을 할 수 있는 CSS를 구성하려면 어떻게 해야 할까요?

jaeyooniverse commented:
![image](https://user-images.githubusercontent.com/103628685/175330411-9f9cfd0d-0abb-4ebb-89fd-e0d62ce48885.png)
![image](https://user-images.githubusercontent.com/103628685/175330567-f402b797-8b96-4e63-825d-d83895db5b6c.png)

6주 차부터 에러 처리에 대한 리뷰를 처음 받고 이런 저런 시도도 많이 해보고 이상적인 에러 핸들링은 무엇일까 혼자 고민도 많이 했는데 도저히 답을 얻을 수 없어서 도움 요청드립니다. 컴포넌트에서나 apu에서 try catch를 다음과 같은 식으로 작성했다고 하면 테스트 코드는 어떤 방식으로 짜는 게 좋을지 도움 부탁드립니다.

jaeyooniverse committed:
전체에 box-sizing: border-box 적용

CSS Box model을 도입하기 위해 전체 style에 box-sizing: border-box 옵션을 적용했습니다.

hannut91 commented:
> 생에 처음으로 CSS를 제대로 하려니까 어떻게 기준을 잡고 시작해야 할 지 너무 막막하네요.
예전 간단한 홈페이지를 만들 때도 UI가 어떻게 그려질지 전혀 예상이 되지 않는 채로 수치를 막 입력한 적이 있었는데,
예측을 할 수 있는 CSS를 구성하려면 어떻게 해야 할까요?

CSS는 일종의 계산식입니다. 계산식에 값을 입력하면 계산된 값으로 화면에 그림을 그립니다. 이 것을 잘 알려면 CSS 속성들을 정확하게 알고 있어야 합니다. 많이 만들어보고 이게 왜 이렇게 화면에 그려지지? 를 정확하게 공부해 보시길 추천드립니다.

See also
  - https://www.youtube.com/watch?v=_o1zsrBkZyg&list=PLBNdLLaRx_rKXwi7MulM6v1UG9JLKWIYS

hannut91 commented:
일단 컴포넌트에서는 try catch문을 사용하는 것은 아닙니다. 예외가 발생하고 예외가 발생했을 때 처리하기 위해서 try catch문을 사용하는 것인데요. 위의 컴포넌트 코드에서는 예외를 던지지 않습니다.

Flux 아키텍처에 대해서 다시 한 번 보세요. 상태가 있고, 이 상태를 보여주는 기능 그리고 화면에서 사용자의 이벤트를 처리하는 것 React입니다. 리액트는 상태에 따라서 화면에 보여주는 일만 한다고 생각해 보세요. 어떤 상태가 됐으니 무언가를 보여주는 겁니다.
어떤 에러가 난 상태가 되었더니 사용자에게 무엇을 보여주는거라고 생각해 보세요. 그러면 상태를 바꾸는 것과 보여주는 것 두 가지가 나뉘어 지게 됩니다.

https://facebook.github.io/flux/docs/in-depth-overview/
